kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+    final HashMap<String, Object> configs = new HashMap<>(providedConfigs);
+    configs.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    configs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    return Map.copyOf(configs);
+  }
+
+  public ResponsiveConsumer.Listener addForThread(final String threadId) {
+    LOGGER.info("add end offset metrics for thread {}", threadId);
+    final var tm = new ThreadMetrics(threadId);
+    synchronized (this) {
+      assert !threadIdToMetrics.containsKey(threadId);
+      if (threadIdToMetrics.isEmpty()) {
+        initPoller();
+        threadIdToMetrics.put(threadId, tm);
+      }
+    }
+    return tm;
+  }
+
+  public void removeForThread(final String threadId) {
+    final ThreadMetrics tm;
+    synchronized (this) {
+      tm = threadIdToMetrics.remove(threadId);
+      if (tm != null) {
+        tm.cleanup();
+        if (threadIdToMetrics.isEmpty()) {
+          stopPoller();
+        }
+      }
+    }
+  }
+
+  private void pollEndOffsets() {
+    final Set<TopicPartition> partitions = new HashSet<>();
+    final Collection<ThreadMetrics> threadMetrics;
+    synchronized (this) {
+      threadMetrics = List.copyOf(this.threadIdToMetrics.values());
+    }
+    for (final var tm : threadMetrics) {
+      partitions.addAll(tm.endOffsets.keySet());
+    }
+    final var endOffsets = consumer.endOffsets(partitions);
+    threadMetrics.forEach(tm -> tm.update(endOffsets));
+  }
+
+  private void initPoller() {
+    LOGGER.info("init end offsets poller");
+    assert consumer == null;
+    assert executor == null;
+    consumer = factories.createConsumer(configs);
+    executor = factories.createExecutor();
+    pollFuture = executor.scheduleAtFixedRate(this::pollEndOffsets, 0, 30, TimeUnit.SECONDS);
+  }
+
+  private void stopPoller() {
+    LOGGER.info("stopping end offsets poller");
+    try {
+      consumer.close(Duration.ofNanos(0));
+    } catch (final Exception e) {
+      LOGGER.warn("error closing consumer", e);
+    }
+    consumer = null;
+    pollFuture.cancel(true);
+    executor.shutdown();
+    executor = null;
+  }
+
+  private class ThreadMetrics implements ResponsiveConsumer.Listener {
+    private final String threadId;
+    private final Map<TopicPartition, Long> endOffsets = new ConcurrentHashMap<>();
+
+    private ThreadMetrics(final String threadId) {
+      this.threadId = Objects.requireNonNull(threadId);
+    }
+
+    @Override
+    public void onPartitionsLost(final Collection<TopicPartition> lost) {
+      onPartitionsRevoked(lost);
+    }
+
+    @Override
+    public void onPartitionsRevoked(final Collection<TopicPartition> revoked) {
+      for (final var p : revoked) {
+        metrics.removeMetric(metricName(p));
+        endOffsets.remove(p);
+      }
+    }
+
+    @Override
+    public void onPartitionsAssigned(final Collection<TopicPartition> assigned) {
```

Comment:
I believe @ableegoldman confirmed with me that this is only called after the first `poll` (you don't get assigned any partitions on an assignment until you poll) - I believe we'd never get anything assigned to us?

Reply:
> I believe we'd never get anything assigned to us?

I'm not following - when the stream thread calls `poll` and an assignment happens - this will get called right?

Reply:
I think I may have misread the code, but I thought we were using a consumer that we created just for the `EndOffsetsPoller` - I can't find it now though so we can ignore that.